lte/gateway/c/core/oai/tasks/amf/amf_app_state_converter.cpp

```diff
@@ -0,0 +1,240 @@
+    }
+  }
+}
+
+// Converts Proto to Map<guti_m5_t,uint64_t>
+void AmfNasStateConverter::proto_to_guti_map(
+    const google::protobuf::Map<std::string, uint64_t>& proto_map,
+    map_guti_m5_uint64_t* guti_map) {
+  for (auto const& kv : proto_map) {
+    amf_ue_ngap_id_t amf_ue_ngap_id = kv.second;
+    std::unique_ptr<guti_m5_t> guti = std::make_unique<guti_m5_t>();
+    memset(guti.get(), 0, sizeof(guti_m5_t));
+    // Converts guti to string.
+    amf_app_convert_string_to_guti_m5(guti.get(), kv.first);
+
+    guti_m5_t guti_received = *guti.get();
+    magma::map_rc_t m_rc    = guti_map->insert(guti_received, amf_ue_ngap_id);
+    if (m_rc != magma::MAP_OK) {
+      OAILOG_ERROR(
+          LOG_AMF_APP,
+          "Failed to insert amf_ue_ngap_id %lu in GUTI table, error: %s\n",
+          amf_ue_ngap_id, map_rc_code2string(m_rc).c_str());
+    }
+  }
+}
+
+/*********************************************************
+ *                AMF app state<-> Proto                  *
+ * Functions to serialize/desearialize AMF app state      *
+ * The caller is responsible for all memory management    *
+ **********************************************************/
+
+void AmfNasStateConverter::state_to_proto(
+    const amf_app_desc_t* amf_nas_state_p,
+    magma::lte::oai::MmeNasState* state_proto) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
+  state_proto->set_mme_app_ue_s1ap_id_generator(
+      amf_nas_state_p->amf_app_ue_ngap_id_generator);
+
+  // maps to proto
+  auto amf_ue_ctxts_proto = state_proto->mutable_mme_ue_contexts();
+  OAILOG_DEBUG(LOG_AMF_APP, "IMSI table to proto");
+  map_uint64_uint64_to_proto(
+      amf_nas_state_p->amf_ue_contexts.imsi_amf_ue_id_htbl,
+      amf_ue_ctxts_proto->mutable_imsi_ue_id_htbl());
+  map_uint64_uint64_to_proto(
+      amf_nas_state_p->amf_ue_contexts.tun11_ue_context_htbl,
+      amf_ue_ctxts_proto->mutable_tun11_ue_id_htbl());
+  map_uint64_uint64_to_proto(
+      amf_nas_state_p->amf_ue_contexts.gnb_ue_ngap_id_ue_context_htbl,
+      amf_ue_ctxts_proto->mutable_enb_ue_id_ue_id_htbl());
+  map_guti_uint64_to_proto(
+      amf_nas_state_p->amf_ue_contexts.guti_ue_context_htbl,
+      amf_ue_ctxts_proto->mutable_guti_ue_id_htbl());
+  OAILOG_FUNC_OUT(LOG_AMF_APP);
+}
+
+void AmfNasStateConverter::proto_to_state(
+    const magma::lte::oai::MmeNasState& state_proto,
+    amf_app_desc_t* amf_nas_state_p) {
+  OAILOG_FUNC_IN(LOG_AMF_APP);
+  amf_nas_state_p->amf_app_ue_ngap_id_generator =
+      state_proto.mme_app_ue_s1ap_id_generator();
+
+  if (amf_nas_state_p->amf_app_ue_ngap_id_generator == 0) {  // uninitialized
+    amf_nas_state_p->amf_app_ue_ngap_id_generator = 1;
+  }
+  OAILOG_INFO(LOG_AMF_APP, "Done reading AMF statistics from data store");
+
+  magma::lte::oai::MmeUeContext amf_ue_ctxts_proto =
+      state_proto.mme_ue_contexts();
+
+  amf_ue_context_t* amf_ue_ctxt_state = &amf_nas_state_p->amf_ue_contexts;
+
+  // proto to maps
+  OAILOG_INFO(LOG_AMF_APP, "Hashtable AMF UE ID => IMSI");
```

Comment:
@GANESH-WAVELABS is the action for these to be removed later given the AMF stateless improvements plan? If so @Krishnavamsi-wavelabs can you please mark it with a comment?

Reply:
-  Even after replacing maps defined in amf_ue_context_t to google::protobuf::Map, amf has to copy  google::protobuf::Maps  from amf_app_desc_t  to  state_proto(proto data structure).  I thought I can avoid copy after changing to google::protobuf::Map.

-   It was major changes in amf application code, I wanted to remove these all maps as part of the enhancement activity proposed.
Please provide your comment

Reply:
@ganeshg87 the idea would be to use the proto data structures to be the local hashtables instead, to avoid copying each member of the local hashtable to the proto structure. That should be the way to go going forward also for the ue_id to ue_context hashtables, instead of using `std::unordered_map` we can use the `google::protobuf::Map` protobuf object instead (on `lte/gateway/c/core/oai/include/map.h` class)